pymc_marketing/mmm/budget_optimizer.py

```diff
@@ -0,0 +1,171 @@
+    """
+    Optimize the budget allocation across channels to maximize total contribution.
+
+    Using the Michaelis-Menten model, this function seeks the best budget distribution across
+    channels that maximizes the total expected contribution.
+
+    This function leverages the Sequential Least Squares Quadratic Programming (SLSQP) optimization
+    algorithm to find the best budget distribution across channels that maximizes the total
+    expected contribution based on the Michaelis-Menten model.
+
+    The optimization is constrained such that:
+    1. The sum of budgets across all channels equals the total available budget.
+    2. The budget allocated to each individual channel lies within its specified range.
+
+    The SLSQP method is particularly suited for this kind of problem as it can handle
+    both equality and inequality constraints.
+
+    Parameters
+    ----------
+    total_budget : int
+        The total budget to be distributed across channels.
+    budget_ranges : Dict or None
+        An optional dictionary defining the minimum and maximum budget for each channel.
+        If not provided, the budget for each channel is constrained between 0 and its L value.
+    parameters : Dict
+        Michaelis-Menten parameters for each channel as described in `calculate_expected_contribution`.
+    channels : list of str
+        The list of channels for which the budget is being optimized.
+
+    Returns
+    -------
+    Dict
+        A dictionary with channels as keys and the optimal budget for each channel as values.
+    """
+
+    # Check if budget_ranges is the correct type
+    if not isinstance(budget_ranges, (dict, type(None))):
+        raise TypeError("`budget_ranges` should be a dictionary or None.")
+
+    if budget_ranges is None:
+        budget_ranges = {
+            channel: (0, min(total_budget, parameters[channel][0]))
+            for channel in channels
+        }
+
+    initial_guess = [total_budget // len(channels)] * len(channels)
+
+    bounds = [budget_ranges[channel] for channel in channels]
+
+    constraints = {"type": "eq", "fun": lambda x: np.sum(x) - total_budget}
+
+    result = minimize(
+        objective_distribution,
+        initial_guess,
+        args=(channels, parameters),
+        method="SLSQP",
+        bounds=bounds,
+        constraints=constraints,
+    )
+
+    return {channel: budget for channel, budget in zip(channels, result.x)}
+
+
+def budget_allocator(
+    total_budget: int,
+    channels: List[str],
+    parameters: Dict[str, Tuple[float, float]],
+    budget_ranges: Optional[Dict[str, Tuple[float, float]]],
+) -> DataFrame:
+
+    optimal_budget = optimize_budget_distribution(
+        total_budget, budget_ranges, parameters, channels
+    )
+
+    return DataFrame(
+        {
+            "estimated_contribution": calculate_expected_contribution(
+                parameters, optimal_budget
+            ),
+            "optimal_budget": optimal_budget,
```

Comment:
From the example notebook I get 
![image](https://github.com/pymc-labs/pymc-marketing/assets/22996444/f6cc821c-5f9f-41b2-aff8-85714594e292)
Should we add the expected total value instead of `NaN`?

Reply:
Correct, already working as expected.

<img width="626" alt="Screenshot 2023-08-29 at 00 40 31" src="https://github.com/pymc-labs/pymc-marketing/assets/59846724/a68191ad-4a80-407f-9040-575e4f503727">